### PR TITLE
Stopped forwarding TypeScript diagnostics

### LINKED ISSUE #48: Does not run on .d.ts declaration files
### 🐛 Bug Report

- `eslint-plugin-expect-type` version: 0.1.0
- ESLint version: N/A
- Node version: N/A

#### Actual Behavior

This paraphrasing of code in `expect.ts` explicitly stops this plugin from operating on declaration files:

```ts
  if (sourceFile.isDeclarationFile || !/\$Expect(Type|Error)/.test(sourceFile.text)) {
    // ...
    return;
  }
```

#### Expected Behavior

Users are going to want to run the assertions inside `.d.ts` files. It might have been intentional originally, but I see this as a bug.

The `sourceFile.isDeclarationFile` check should be removed.

#### Reproduction

```ts
// index.d.ts
// $ExpectType number
export const value: string;
```

### LINKED ISSUE #49: Stop forwarding TypeScript diagnostics
### 🐛 Bug Report

- `eslint-plugin-expect-type` version: 0.1.0
- ESLint version: N/A
- Node version: N/A

#### Actual Behavior

Right now, any pre-emit diagnostic (so, some classes of compile errors) are emitted in `expect.ts`. See the `addDiagnosticFailure` function.

```ts
  const diagnostics = ts.getPreEmitDiagnostics(program, sourceFile);

  // ...

  for (const diagnostic of diagnostics) {
    addDiagnosticFailure(diagnostic);
  }
```

#### Expected Behavior

There's no need for the tool to do this. I see it as a bug, not a feature.

The `addDiagnosticFailure` function should be removed.

#### Reproduction

```ts
export const duplicate: number;
export const duplicate: number;
```

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

- [x] Addresses an existing issue: fixes #48; fixes #49
- [x] That issue was marked as [`status: accepting prs`](https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type/labels/status%3A%20accepting%20prs)

## Overview

#49: Removes the `TypeScriptCompileError` error message ID and `addDiagnosticFailure` function that created it. Moves the potentially expensive `ts.getPreEmitDiagnostics()` call to after the _"does this file have any assertions?"_ quick check.

#48: Since it's the same area, removed the `sourceFile.isDeclarationFile` check too.

Neither of those behaviors were tested, so this should increase overall test coverage. Yay!

### CODE REVIEW

File: src/rules/expect.ts [hunk 1]
```diff
@@ -123,15 +117,7 @@ function validate(context: TSESLint.RuleContext<MessageIds, [Options]>, options:
...
-  const seenDiagnosticsOnLine = new Set<number>();
-
-  for (const diagnostic of diagnostics) {
-    const line = lineOfPosition(diagnostic.start!, sourceFile);
-    seenDiagnosticsOnLine.add(line);
-    if (!errorLines.has(line)) {
-      addDiagnosticFailure(diagnostic);
-    }
-  }
+  const seenDiagnosticsOnLine = new Set(diagnostics.map((diagnostic) => lineOfPosition(diagnostic.start!, sourceFile)));
```
Comment (danvk) [inline src/rules/expect.ts:120]: Do you know why `diagnostic.start` is nullable / optional? Add a comment about why this is safe.

**Review by danvk — APPROVED** [review]
Nice!

Speaking of code coverage, it would be nice to set up something that tracks it for this repo. (I recently did that with codecov on [crudely-typed](https://github.com/danvk/crudely-typed) and am happy with the results). The GitHub action logs reports 82.59% statement coverage, which does seem kinda low for an eslint rule.

File: src/rules/expect.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [reply src/rules/expect.ts:120]: I honestly have no idea.

File: src/rules/expect.ts [hunk 2]
```diff
@@ -99,17 +98,12 @@ function validate(context: TSESLint.RuleContext<MessageIds, [Options]>, options:
     return;
   }
 
-  const checker = program.getTypeChecker();
-  // Don't care about emit errors.
-  const diagnostics = ts.getPreEmitDiagnostics(program, sourceFile);
-  if (sourceFile.isDeclarationFile || !/(?:\$Expect(Type|Error|^\?))|\^\?/.test(sourceFile.text)) {
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply src/rules/expect.ts:90]: Not yet, but I'll work on it 🙃 https://github.com/typescript-eslint/typescript-eslint/issues/4845

[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply src/rules/expect.ts:90]: Looks like I'm giving up on that effort. I'll file an issue to revisit.

[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply src/rules/expect.ts:90]: #55

**Comment by JoshuaKGoldberg** [comment]
#56 for tracking coverage 👍

**Pull request merged**